### DISCUSSION TITLE
Struggling to install Firedrake on Mac

### DISCUSSION THREAD

**Discussion by richard-parsons** [firedrakeproject/firedrake#3671] [Firedrake support]
I've been trying to install firedrake on mac but without success. I've still not been able to install successfully after going through suggestions from past discussions and the install FAQ. 

I've attached the install log and configure log (using latest version of MacOS and xcode), any suggestions would be greatly appreciated! 
Thanks.

[firedrake-install.log](https://github.com/user-attachments/files/16118746/firedrake-install.log)
[configure.log](https://github.com/user-attachments/files/16118747/configure.log)

**Comment by dham** [comment]
The error seems to be saying that your linker is somehow incompatible with your compiler. I'm not quite sure why that would be.

**Comment by Ig-dolci** [comment]
Could you please try this [firedrake_install script](https://github.com/firedrakeproject/firedrake/blob/dolci/remove_ld_classic/scripts/firedrake-install)?

**Reply by richard-parsons** [reply]
Thanks for your help, I tried that script and got the same error message. I've attached the log files again incase there's any other helpful info in there. 

[firedrake-install.log](https://github.com/user-attachments/files/16126205/firedrake-install.log)
[configure.log](https://github.com/user-attachments/files/16126206/configure.log)

Also - and I'm not sure if this was a sensible thing to try - I tried installing using a previous release of xcode and the build seemed to get a bit further and exited with the error message failing to install openblas. I tried the suggested script updates for this as per the FAQ but the error message didn't change.

**Reply by Ig-dolci** [reply]
Ahh. I just observed your python installer is `/usr/local/opt/python@3.12/bin/python3.12` (perhaps not a brew python). Can you try again with `$(brew --prefix)/bin/python3 firedrake-install` (firedrake-install script from `curl -O https://raw.githubusercontent.com/firedrakeproject/firedrake/master/scripts/firedrake-install`)?

**Reply by connorjward** [reply]
> Ahh. I just observed your python installer is `/usr/local/opt/python@3.12/bin/python3.12` (perhaps not a brew python). Can you try again with `$(brew --prefix)/bin/python3 firedrake-install` (firedrake-install script from `curl -O https://raw.githubusercontent.com/firedrakeproject/firedrake/master/scripts/firedrake-install`)?

I think this probably is a valid homebrew path. What does `brew --prefix` show?

**Reply by richard-parsons** [reply]
I had been using $(brew --prefix)/bin/python3 firedrake-install.
brew --prefix shows /usr/local

**Reply by connorjward** [reply]
I notice that `brew doctor` is throwing up a number of warnings. Would you be able to address those?

**Reply by richard-parsons** [reply]
I've addressed the warnings from brew doctor, but I was still getting the same error message as before.

I changed the MacOS development target from 12.0 to 14.0 (not sure if this is sensible?) and the build got further, exiting with 'downloaded openblas could not be used'. I tried the script update from the FAQs for the openblas issue but with no luck. Log files attached again, thanks.

[firedrake-install.log](https://github.com/user-attachments/files/16130780/firedrake-install.log)
[configure.log](https://github.com/user-attachments/files/16130791/configure.log)

**Reply by connorjward** [reply]
I see lines like:
```
ld: warning: ignoring file /Users/richardparsons/firedrake/src/petsc/default/lib/libopenblas.dylib, building for macOS-x86_64 but attempting to link with file built for unknown-unsupported file format ( 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 )
```
I have found [this post online](https://stackoverflow.com/questions/70126695/ld-building-for-macos-x86-64-but-attempting-to-link-with-file-built-for-macos-x) that suggests that you are experiencing clashes between different installed compiler toolchains. Could you try following the instructions in the top solution there?

**Reply by richard-parsons** [reply]
Thanks for all of your help and suggestions with this, I have managed to complete the installation now (I think). I checked the instructions from the attached post above and that seemed to be ok.

To get the install to run, I updated the MACOSX_DEPLOYMENT_TARGET to one that matched my current mac os version, which got past the linker issue.
I then updated line 780 of the install script but also found I needed to set 'export BLAS=/Users/richardparsons/firedrake/src/petsc/default/lib/libopenblas.dylib' to point to the openblas installation to get past that issue.

I don't really understand what these changes are doing and they may well be contributing to my next problem which is running the recommended pytests to check the installation which return a long script of errors. The top of this reads:


tests/regression/test_dg_advection.py FFFF                               [ 14%]
tests/regression/test_poisson_strong_bcs.py FFFFFFFFFFFFFFFFF            [ 77%]
tests/regression/test_poisson_strong_bcs_nitsche.py FFFF                 [ 92%]
tests/regression/test_stokes_mini.py FF                                  [100%]

=================================== FAILURES ===================================
_____________________ test_dg_advection_icosahedral_sphere _____________________

name = 'siphash13', data = b'', kwargs = {}

    def __hash_new(name, data=b'', **kwargs):
        """new(name, data=b'') - Return a new hashing object using the named algorithm;
        optionally initialized with data (which must be a bytes-like object).
        """
        if name in __block_openssl_constructor:
            # Prefer our builtin blake2 implementation.
            return __get_builtin_constructor(name)(data, **kwargs)
        try:
>           return _hashlib.new(name, data, **kwargs)
E           _hashlib.UnsupportedDigestmodError: [digital envelope routines] unsupported

/usr/local/Cellar/python@3.12/3.12.4/Frameworks/Python.framework/Versions/3.12/lib/python3.12/hashlib.py:160: UnsupportedDigestmodError


Also 'ValueError: unsupported hash type siphash13' seems to come up frequently in a long chain of error messages.

Have you seen these errors before or have any ideas what the issue might be?

**Reply by dham** [reply]
Ah, this is bad timing. One of our upstream dependencies, `pytools` put out a change that required us to update `loopy` in response. This meant that our build was out for about 48 hours from Tues to Thurs this week. It should be fixed now so simply run the install again.

If you are interested, there is a bit more discussion about this on our Slack channel (though there is no real reason to care: the key thing is that it should now be fixed).

**Reply by richard-parsons** [reply]
Thanks a lot, I've got it installed now and tests have completed successfully.
Thanks for all your help, its very much appreciated!
